provider/lxd/server.go

```diff
@@ -0,0 +1,321 @@
+	DeleteCertificate(fingerprint string) (err error)
+	CreateClientCertificate(certificate *lxd.Certificate) error
+	LocalBridgeName() string
+	AliveContainers(prefix string) ([]lxd.Container, error)
+	ContainerAddresses(name string) ([]network.Address, error)
+	RemoveContainer(name string) error
+	RemoveContainers(names []string) error
+	FilterContainers(prefix string, statuses ...string) ([]lxd.Container, error)
+	CreateContainerFromSpec(spec lxd.ContainerSpec) (*lxd.Container, error)
+	WriteContainer(*lxd.Container) error
+	CreateProfileWithConfig(string, map[string]string) error
+	GetProfile(string) (*lxdapi.Profile, string, error)
+	HasProfile(string) (bool, error)
+	VerifyNetworkDevice(*lxdapi.Profile, string) error
+	StorageSupported() bool
+	GetStoragePool(name string) (pool *lxdapi.StoragePool, ETag string, err error)
+	GetStoragePools() (pools []lxdapi.StoragePool, err error)
+	CreatePool(name, driver string, attrs map[string]string) error
+	GetStoragePoolVolume(pool string, volType string, name string) (*lxdapi.StorageVolume, string, error)
+	GetStoragePoolVolumes(pool string) (volumes []lxdapi.StorageVolume, err error)
+	CreateVolume(pool, name string, config map[string]string) error
+	UpdateStoragePoolVolume(pool string, volType string, name string, volume lxdapi.StorageVolumePut, ETag string) error
+	DeleteStoragePoolVolume(pool string, volType string, name string) (err error)
+	ServerCertificate() string
+	EnableHTTPSListener() error
+}
+
+// ServerFactory creates a new factory for creating servers that are required
+// by the server.
+type ServerFactory interface {
+	// LocalServer creates a new lxd server and augments and wraps the lxd
+	// server, by ensuring sane defaults exist with network, storage.
+	LocalServer() (Server, error)
+
+	// LocalServerHostName returns the local servers host name from the factory.
+	LocalServerHostName() (string, error)
+
+	// RemoteServer creates a new server that connects to a remote lxd server.
+	// If the cloudSpec endpoint is nil or empty, it will assume that you want
+	// to connection to a local server and will instead use that one.
+	RemoteServer(environs.CloudSpec) (Server, error)
+}
+
+// InterfaceAddress groups methods that is required to find addresses
+// for a given interface
+type InterfaceAddress interface {
+
+	// InterfaceAddress looks for the network interface
+	// and returns the IPv4 address from the possible addresses.
+	// Returns an error if there is an issue locating the interface name or
+	// the address associated with it.
+	InterfaceAddress(string) (string, error)
+}
+
+type interfaceAddress struct{}
+
+func (interfaceAddress) InterfaceAddress(interfaceName string) (string, error) {
+	return utils.GetAddressForInterface(interfaceName)
+}
+
+type localServerFunc func() (Server, error)
+type remoteServerFunc func(lxd.ServerSpec) (Server, error)
+
+type serverFactory struct {
+	newLocalServerFunc  localServerFunc
```

Comment:
We already indirect the factory with ```ServerFactory```. We can eschew these funcs.
Just use export_test to get at ```bootstrapLocalServer```.